File: ruby/anagrams/anagrams.rb

def anagramRotate(chars, at)
  length = chars.size
  return if (at >= length)

  temp = chars[at]
  (at..(length-1)).each { |i|
    chars[i] = chars[i + 1]
  }
  chars[length - 1] = temp
end


def anagramRecurse(chars, at)
  chars_left = chars.size - at
  return if (chars_left == 1)

  (0...chars_left).each { |i|
    anagramRecurse(chars, at + 1)
    if (chars_left == 2)
      puts chars.join('')
    end
    anagramRotate(chars, at)
  }
end

def anagram(text)
  chars = text.split('')
  anagramRecurse(chars, 0)
end

print "anagrams for text > "
input = gets.chomp
anagram(input)

File: javascript/anagrams/anagrams.js
const readline = require('readline');

const readlineInterface = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

function anagramRotate (chars, at) {
  if (at >= chars.length) {
    return;
  }

  const temp = chars[at];
  for (let i = at; i < chars.length - 1; i++) {
    chars[i] = chars[i + 1];
  }
  chars[chars.length - 1] = temp;
}

function anagramRecurse (chars, at) {
  const end = chars.length - at;
  if (end == 1) {
    return;
  }

  for (let i = 0; i < end; i++) {
    anagramRecurse(chars, at + 1);
    if (end == 2) {
      console.log(chars.join(''));
    }
    anagramRotate(chars, at);
  }
}

function anagrams (input) {
  let chars = input.split('');
  anagramRecurse(chars, 0);
}

readlineInterface.question('Input a string to anagram > ', (input) => {
  anagrams(input);
  readlineInterface.close();
});

